lib/eslint.js

```diff
@@ -868,10 +870,11 @@ module.exports = (function() {
                         ruleCreator(ruleContext);
 
                     // add all the node types as listeners
-                    Object.keys(rule).forEach(nodeType => {
-                        api.on(nodeType, timing.enabled
-                            ? timing.time(key, rule[nodeType])
-                            : rule[nodeType]
+                    Object.keys(rule).forEach(selector => {
+                        selectors.add(selector);
+                        api.on(selector, timing.enabled
```

Comment:
This is a valid point, but I'm don't see how adding every selector could be avoided. (I noticed a similar discussion on it [here](https://github.com/eslint/eslint/issues/5407).)

As far as I can tell, there are two possible approaches to emitting selectors:

* Check which selectors are needed by rules, and only match against those
* Emit *every* selector that matches a node while traversing, and rules will fire their listeners only for the selectors they care about. This would also result in a very large number of selectors. For example, see the following code:
  ```js
  foo + bar;
  ```
  When traversing, we would need to emit the following selectors for the `foo` node:

  * `Identifier`
  * `BinaryExpression > Identifier`
  * `ExpressionStatement > BinaryExpression > Identifier`
  * `Program > ExpressionStatement > BinaryExpression > Identifier`
  * `BinaryExpression Identifier`
  * `ExpressionStatement Identifier`
  * `ExpressionStatement > BinaryExpression Identifier`
  * `Program Identifier`
  * `:not(ReturnStatement) > Identifier`
  * (a lot more)

I realize that this list would be reduced depending on what selector syntax we support, but it seems like a bad idea to use an approach that would require us to make our selectors less powerful.